### ISSUE TITLE
Syntax error for class method receiving name from method def

### ISSUE THREAD

**Issue by brixen** [rubinius/rubinius-code#7]
_Unsure if this is still an issue. Reported by @jemc on rubinius-melbourne. Moving here._

The most recent version of melbourne has a syntax incompatibility compared to MRI when passing the name of a method returned by a method `def` as an argument to a class method.  It only occurs when there is a call within the method that has a block using the `do`/`end` syntax - the `{`/`}` syntax is okay.  Here is a minimal reproducible example and its behavior with both Rubinius and MRI:

``` bash
 ruby --version
rubinius 2.5.5.c51 (2.1.0 f2116066 2015-06-07 3.5.0 JI) [x86_64-linux-gnu]

 ruby -e "p Rubinius::ToolSets::Runtime::Melbourne::VERSION"
"2.3.1.0"

 ruby -c -e "
private def some_method
  some_enumerable.each do |item|
    item
  end
end
"
-e:2: expecting keyword_end: -e:2:37
-e:5: expecting $end: -e:5:15

 ruby -c -e "
private def some_method
  some_enumerable.each { |item|
    item
  }
end
"
Syntax OK

 ruby -c -e "
def some_method
  some_enumerable.each do |item|
    item
  end
end
private :some_method
"
Syntax OK
```

```
 ruby --version
ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]

 ruby -c -e "
private def some_method
  some_enumerable.each do |item|
    item
  end
end
Syntax OK

 ruby -c -e "
private def some_method
  some_enumerable.each { |item|
    item
  }
end
"
Syntax OK

 ruby -c -e "
def some_method
  some_enumerable.each do |item|
    item
  end
end
private :some_method
"
Syntax OK
```

**Comment by jemc** [comment]
Just checked - still an issue with:

``` ruby
 ruby --version
rubinius 3.15.c80 (2.2.2 aac99d56 2016-02-21 3.5.0 JI) [x86_64-linux-gnu]

 ruby -e "p Rubinius::ToolSets::Runtime::Melbourne::VERSION"
"3.3"
```

**Comment by andrew-aladev** [comment]
+1 it is still an issue with "private def" and "protected def".

**Comment by brixen** [comment]
@andrew-aladev please paste the output of `ruby -v` for Rubinius.

This is what I have with 3.73:

```
$  ruby -v -c -e "
private def some_method
  some_enumerable.each { |item|
    item
  }
end
"
rubinius 3.73.c1 (2.3.1 4d6a87d5 2017-04-01 3.8.1) [x86_64-darwin16.4.0]
Syntax OK
```

**Comment by andrew-aladev** [comment]
@brixen, Hello. I am running latest rubinius 3.73.
```
ruby -v -c -e "protected def a; b.each do; end; end"
rubinius 3.73 (2.3.1 45c1bd27 2017-03-31 3.9.1) [x86_64-linux-gnu]
-e:1: expecting keyword_end: -e:1:26
```

We can reproduce the same issue with any block.

```
protected def a
  Proc.new do; end
end
```

**Comment by brixen** [comment]
@andrew-aladev thanks for the clarification. This appears to be a different issue than the original reproduction, despite looking similar.

The minimal code to reproduce is:

```
$ ruby -v -c -e "protected def a; a.each do; end; end"
rubinius 3.73.c1 (2.3.1 4d6a87d5 2017-04-01 3.8.1) [x86_64-darwin16.4.0]
-e:1: expecting keyword_end: -e:1:26
```

Both of the following parse correctly:

```
$ ruby -v -c -e "protected def a; a do; end; end"
rubinius 3.73.c1 (2.3.1 4d6a87d5 2017-04-01 3.8.1) [x86_64-darwin16.4.0]
Syntax OK
```

```
$ ruby -v -c -e "protected def a; a.each {}; end"
rubinius 3.73.c1 (2.3.1 4d6a87d5 2017-04-01 3.8.1) [x86_64-darwin16.4.0]
Syntax OK
```

**Comment by brixen** [comment]
@andrew-aladev ignore what I said about it being a different issue, I misread the original post. The minimal repro in my comment above is correct.

**Comment by brixen** [comment]
This is an even more minimal repro:

```
ruby -v -c -e "x def a; a.b do; end; end"
```

**Comment by brixen** [comment]
@andrew-aladev this is fixed in 3.74. For reference, the issue was https://bugs.ruby-lang.org/issues/9308. In Melbourne, the table is in a separate header and this change isn't visible in the Ruby grammar itself, so I missed it.

```
$ ruby -v -c -e "x def a; a.b do; end; end"
rubinius 3.74 (2.3.1 5d398da1 2017-04-16 3.8.1) [x86_64-darwin16.5.0]
Syntax OK
```

**Comment by andrew-aladev** [comment]
Ok. thank you.
